azure/durable_functions/tasks/task_utilities.py

```diff
@@ -191,3 +192,208 @@ def set_processed(tasks):
+    scheduled_task: Optional[HistoryEventType]:
+        The sub-orchestration creation event, if found.
+
+    Returns
+    -------
+    Optional[HistoryEventType]:
+        The matching sub-orchestration completed event, if found. Else, None.
+    """
+    event_type = HistoryEventType.SUB_ORCHESTRATION_INSTANCE_COMPLETED
+    return find_sub_orchestration(
+        state=state,
+        event_type=event_type,
+        scheduled_task=scheduled_task)
+
+
+def find_sub_orchestration_failed(
+        state: List[HistoryEvent],
+        scheduled_task: Optional[HistoryEvent]) -> Optional[HistoryEvent]:
+    """Look-up the sub-orchestration failure event.
+
+    Parameters
+    ----------
+    state: List[HistoryEvent]:
+        The history of Durable events
+    scheduled_task: Optional[HistoryEvent]:
+        The sub-orchestration creation event, if found.
+
+    Returns
+    -------
+    Optional[HistoryEvent]:
+        The matching sub-orchestration failure event, if found. Else, None.
+    """
+    event_type = HistoryEventType.SUB_ORCHESTRATION_INSTANCE_FAILED
+    return find_sub_orchestration(
+        state=state,
+        event_type=event_type,
+        scheduled_task=scheduled_task)
+
+
+def find_matching_event(
+        state: List[HistoryEvent],
+        event_type: HistoryEventType,
+        scheduled_task: Optional[HistoryEvent] = None) -> Optional[HistoryEvent]:
+    """Find matching event in the state array, if it exists.
+
+    Parameters
+    ----------
+    state: List[HistoryEvent]:
+        The list of Durable events
+    event_type: HistoryEventType:
+        The type of event being searched-for.
+    scheduled_task" Optional[HistoryEvent], optional:
+        The corresponding `scheduled` task for the searched-for event,
+        only available when looking for a completed or failed event.
+        Defaults to None.
+
+    Returns
+    -------
+    Optional[HistoryEvent]:
+        The matching event from the state array, if it exists.
+    """
+
+    def should_preserve(event: HistoryEvent) -> bool:
+        """Check if `event` matches the task being searched-for.
+
+        Parameters
+        ----------
+        event: HistoryEvent:
+            An event from the `state` array.
+
+        Returns
+        -------
+        bool:
+            True if `event` matches the task being search-for.
+            False otherwise.
+        """
+        has_correct_type = event.event_type == event_type
+        is_not_processed = not event.is_processed
+        if not (scheduled_task is None):
+            extra_constraints = event.TaskScheduledId == scheduled_task.event_id
+        should_preserve = has_correct_type and is_not_processed and extra_constraints
+        return should_preserve
+
+    event: Optional[HistoryEvent] = None
+
+    # Preverse only the elements of the state array that correspond with the looked-up event
+    matches = filter(should_preserve, state)
+    matches = list(matches)
+
+    # if len(matches) > 1:
```

Comment:
Let's make sure we actually implement the error handling here.

Reply:
Actually, we did not need that error handling, that was based on an incorrect assumption. Turns out that matches may have a length greater than one at many points, but that we always choose the first element of the array as that's the first non-processed element, which we assume corresponds to our lookup. This is due for a refactor in the future but that would be for a different PR, since it would affect other task-lookup procedures